src/Microsoft.VisualStudio.LanguageServices.Razor/Editor/RazorDirectiveCompletionProvider.cs

```diff
@@ -0,0 +1,185 @@
+            {
+                // Not a Razor file.
+                return Task.FromResult(0);
+            }
+
+            if (!_codeDocumentProvider.TryGetFromDocument(context.Document, out var codeDocument))
+            {
+                // A Razor code document has not yet been associated with the document.
+                return Task.FromResult(0);
+            }
+
+            var syntaxTree = codeDocument.GetSyntaxTree();
+
+            if (!AtDirectiveCompletionPoint(syntaxTree, context))
+            {
+                // Can't have a valid directive at the current location.
+                return Task.FromResult(0);
+            }
+
+            var directives = syntaxTree.Options.Directives.Concat(DefaultDirectives);
+            foreach (var directive in directives)
+            {
+                var propertyDictionary = new Dictionary<string, string>(StringComparer.Ordinal);
+
+                if (!string.IsNullOrEmpty(directive.Description))
+                {
+                    propertyDictionary[DescriptionKey] = directive.Description;
+                }
+
+                var completionItem = CompletionItem.Create(
+                    directive.Directive,
+                    // This groups all Razor directives together
+                    sortText: "_RazorDirective_",
+                    rules: CompletionItemRules.Create(formatOnCommit: false),
+                    tags: ImmutableArray.Create(CompletionTags.Intrinsic),
+                    properties: propertyDictionary.ToImmutableDictionary());
+                context.AddItem(completionItem);
+            }
+
+            return Task.FromResult(0);
+        }
+
+        private bool AtDirectiveCompletionPoint(RazorSyntaxTree syntaxTree, CompletionContext context)
+        {
+            if (TryGetRazorSnapshotPoint(context, out var razorSnapshotPoint))
+            {
+                var change = new SourceChange(razorSnapshotPoint.Position, 0, string.Empty);
+                var owner = syntaxTree.Root.LocateOwner(change);
+                if (owner.ChunkGenerator is ExpressionChunkGenerator &&
+                    owner.Symbols.All(IsDirectiveCompletableSymbol) &&
+                    // Do not provide IntelliSense for explicit expressions. Explicit expressions will usually look like:
+                    // [@] [(] [DateTime.Now] [)]
+                    owner.Parent?.Children.Count > 1 &&
+                    owner.Parent.Children[1] == owner)
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        protected virtual bool TryGetRazorSnapshotPoint(CompletionContext context, out SnapshotPoint snapshotPoint)
+        {
+            snapshotPoint = default(SnapshotPoint);
+
+            if (context.Document.TryGetText(out var sourceText))
+            {
+                var textSnapshot = sourceText.FindCorrespondingEditorTextSnapshot();
```

Comment:
Thought about making all of this VS specific code try-catch => return false on error. We don't have a 100% grasp on what scenarios VS supports here so I'm thinking it might make sense so we don't explode.

Thoughts?

Reply:
let's not do voodoo